src/lib.rs

```diff
@@ -1,74 +1,97 @@
 use futures::executor::block_on;
 use log::{debug, error};
 use tiny_http::{Response, Server, StatusCode};
-use twitch_api2::twitch_oauth2::{
-    tokens::UserTokenBuilder, ClientId, ClientSecret, Scope, UserToken,
+use twitch_oauth2::{
+    client::SurfError,
+    tokens::{errors::UserTokenExchangeError, UserTokenBuilder},
+    ClientId, ClientSecret, Scope, UserToken,
 };
 use url::Url;
 
+/// Errors when
+#[derive(Debug, thiserror::Error)]
+pub enum AuthFlowError {
+    #[error(transparent)]
+    HookError(#[from] HookError),
+    #[error("could not parse url")]
+    UrlParseError(#[from] url::ParseError),
+    #[error("talking with twitch authentication failed")]
+    ExchangeError(#[from] UserTokenExchangeError<SurfError>),
+}
+
+/// Hook errors
+#[derive(Debug, thiserror::Error)]
+pub enum HookError {
+    #[error("when constructing http server")]
+    TinyHttpError(#[source] Box<(dyn std::error::Error + Sync + Send + 'static)>),
+    #[error("could not parse url")]
+    UrlParseError(#[from] url::ParseError),
+    #[error("failed to do IO operation")]
+    IoError(#[from] std::io::Error),
+}
+
 /// Twitch authentication flow.
-pub fn auth_flow(client_id: &str, client_secret: &str, scopes: Option<Vec<Scope>>) -> UserToken {
+pub fn auth_flow(
+    client_id: &str,
+    client_secret: &str,
+    scopes: Option<Vec<Scope>>,
+    redirect_url: &str,
+) -> Result<UserToken, AuthFlowError> {
+    let redirect_url = Url::parse(&redirect_url)?;
     let mut hook = TwitchAuthHook::new(
         String::from(client_id),
         String::from(client_secret),
         scopes,
-        10666,
-    );
-    let (url, csrf) = hook.builder.generate_url();
+        redirect_url,
+    )?;
+    let (url, _) = hook.builder.generate_url();
     println!(
         "To obtain an authentication token, please visit\n{}",
         url.as_str().to_owned()
     );
-    let code = hook.receive_auth_token().unwrap();
+    let (code, state) = hook.receive_auth_token()?;
     let user_token = block_on(async {
         hook.builder
-            .get_user_token(
-                twitch_oauth2::client::surf_http_client,
-                csrf.secret(),
-                &code,
-            )
+            .get_user_token(twitch_oauth2::client::surf_http_client, &state, &code)
             .await
     });
-    user_token.unwrap()
+    user_token.map_err(Into::into)
 }
 
 // Internal implementation.
-struct TwitchAuthHook {
+pub struct TwitchAuthHook {
```

Comment:
Why public? if there is a good reason please change comment at L61

Reply:
If someone wants to present the url to the user in another way, this would allow them to do so.

Some more things should probably be done to facilitate that however